### PR TITLE
[IMP] gmail: allow to upload attachments when logging an email

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Allow to upload the attachments when logging the email on the partner / lead / ticket.

Technical
=========
The attachments are base 64 encoded and then added in the JSON posted on the Odoo endpoint.

The maximum POST request size on the Gmail side is 50 MB. So when the attachments total size exceed 40 MB we just ignore them and show an error message.

N.B.: when creating an email from Gmail, if the sum of the file size is more than 25 MB, they are automatically uploaded on Drive and the links to those files are added in the email.

Links
=====
Task 2545048
See odoo/odoo/pull/71372

### CODE REVIEW

File: gmail/src/models/attachment.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Represent an attachment in an email.
+ */
+export class Attachment {
```
Comment (awa-odoo) [inline gmail/src/models/attachment.ts:4]: Isn't this class a bit redundant with the existing "GmailAttachment" ?

It's basically the same except for a few renames and encoding.

File: gmail/src/models/email.ts [hunk 2]
```diff
@@ -68,4 +70,29 @@ export class Email {
...
+     * Return the list of the attachments in the email.
+     * Done in a getter and not as a property because this object is serialized and
+     * given to the event handler.
+     */
+    getAttachments(): Attachment[] {
+        const message = GmailApp.getMessageById(this.messageId);
+        const gmailAttachments = message.getAttachments();
+        const attachments: Attachment[] = [];
+
+        // 10 MB maximum per file
```
Comment (awa-odoo) [inline gmail/src/models/email.ts:84]: What is enforcing this 10MB limit? Gmail? Odoo?

File: gmail/src/services/log_email.ts [hunk 3]
```diff
@@ -1,17 +1,35 @@
...
+    const MAXIMUM_ATTACHMENTS_SIZE = 40 * 1024 * 1024;
+
+    let totalAttachmentsSize = 0;
+
+    for (const attachment of email.getAttachments()) {
+        totalAttachmentsSize += attachment.bytesSize;
+        if (totalAttachmentsSize < MAXIMUM_ATTACHMENTS_SIZE) {
+            attachments.push([attachment.name, attachment.content]);
+        } else {
+            Logger.log(`The file ${attachment.name} is ignored because you overcome the total attachment size limit.`);
```
Comment (awa-odoo) [inline gmail/src/services/log_email.ts:26]: "overcome" is usually used in another context (when you "overcome" an obstacle or something).
Here "exceed" is generally better / more technical.

(Same thing goes for your comme message, I think I saw a typo "if" -> "is" in there as well)

**Review by awa-odoo — COMMENTED** [review]
Nothing much to see here, I left a few picky comments.

Cheers!

File: gmail/src/models/email.ts (hunk 2 above)
Comment (std-odoo) [reply gmail/src/models/email.ts:84]: To be sure to not exceed the Gmail quota (our HTTP post request can have 50MB maximum)

File: gmail/src/models/attachment.ts (hunk 1 above)
Comment (std-odoo) [reply gmail/src/models/attachment.ts:4]: Ok, I'll remove this model

File: gmail/src/models/email.ts (hunk 2 above)
Comment (std-odoo) [reply gmail/src/models/email.ts:84]: See https://developers.google.com/apps-script/guides/services/quotas#current_limitations

Comment (std-odoo) [reply gmail/src/models/email.ts:84]: But most of the time with modern mail client, the attachments are automatically added in Gdrive / Onedrive when a limit is exceed (25 for gmail, 33mb for onedrive)

screenshot_spect

File: gmail/src/models/error_message.ts [hunk 4]
```diff
@@ -14,6 +14,8 @@ const _ERROR_CODE_MESSAGES: Record<string, string> = {
         "Oops, looks like you have exhausted your free enrichment requests. Please log in to try again.",
     missing_data: "No data found for this email address.",
     unknown: "Something bad happened. Please, try again later.",
+    // Attachment
+    attachments_size_exceeded: "The attachments are ignored because the total size exceed the limit (40 MB).",
```
Comment (awa-odoo) [inline gmail/src/models/error_message.ts:18]: exceeds*

File: gmail/src/services/log_email.ts [hunk 5]
```diff
@@ -1,18 +1,59 @@
 import { postJsonRpc } from "../utils/http";
+import { escapeHtml } from "../utils/html";
 import { URLS } from "../const";
 import { State } from "../models/state";
+import { Email } from "../models/email";
```
Comment (tde-banana-odoo) [inline gmail/src/services/log_email.ts:5]: Not ordering imports kill dogs. Cute little happy yappy dogs.

File: gmail/src/views/partner.ts [hunk 6]
```diff
@@ -20,15 +20,14 @@ function onEnrichCompany(state: State) {
 }
 
 function onLogEmail(state: State) {
```
Comment (tde-banana-odoo) [inline gmail/src/views/partner.ts:22]: Is it a duplicate of onLogEmailPartner ? Can't it be re-used ?

File: gmail/src/models/email.ts [hunk 7]
```diff
@@ -68,4 +70,40 @@ export class Email {
 
         return email;
     }
+
+    /**
+     * Return the list of the attachments in the email.
+     * Done in a getter and not as a property because this object is serialized and
+     * given to the event handler.
+     *
+     * Return null if the total attachment size limit is exceeded so we do not keep big
```
Comment (tde-banana-odoo) [inline gmail/src/models/email.ts:79]: Docstring does not seem up to date ? as it returns [array, error message] .

File: gmail/src/services/log_email.ts [hunk 8]
```diff
@@ -1,18 +1,59 @@
...
+import { _t } from "../services/translation";
+
+/**
+ * Format the email body before sending it to Odoo.
+ * Add error message at the end of the email, fix some CSS issues,...
+ */
+function _formatEmailBody(email: Email, error: ErrorMessage): string {
+    let body = email.body;
+
+    body = "<span>" + _t("From: ") + escapeHtml(email.contactEmail) + "</span><br/><br/>" + body;
```
Comment (tde-banana-odoo) [inline gmail/src/services/log_email.ts:16]: Not an expert, but isn't it some nicer string formatting than doing concatenation ?

File: gmail/src/services/log_email.ts [hunk 9]
```diff
@@ -1,18 +1,59 @@
...
  * Log the given email body in the chatter of the given record.
  */
-export function logEmail(recordId: number, recordModel: string, emailBody: string): boolean {
+export function logEmail(recordId: number, recordModel: string, email: Email): ErrorMessage {
     const url = State.odooServerUrl + URLS.LOG_EMAIL;
     const accessToken = State.accessToken;
 
+    const [attachments, error] = email.getAttachments();
+
+    const body = _formatEmailBody(email, error);
```
Comment (tde-banana-odoo) [inline gmail/src/services/log_email.ts:46]: Linting guy: some unnecessary blank lines here (+ some could be ordered :D )

**Review by tde-banana-odoo — COMMENTED** [review]
Pouic pouic, some questions / comments !

File: gmail/src/views/partner.ts (hunk 6 above)
Comment (std-odoo) [reply gmail/src/views/partner.ts:22]: We could, but it means that we will need to move this event handler in an other file, and currently all event handlers are in the file in which they are called

**Comment by std-odoo** [comment]
@tde-banana-odoo I've made the changes and answered about the event handler :)

File: gmail/src/views/partner.ts (hunk 6 above)
Comment (tde-banana-odoo) [reply gmail/src/views/partner.ts:22]: That's ok then. Keep it like this :) .

**Pull request merged**